cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java

```diff
@@ -578,10 +582,60 @@ private Future<Void> pauseResume(Reconciliation reconciliation, String host, Kaf
         }
     }
 
+    private  Future<ConnectorStatusAndConditions> autoRestartFailedConnectorAndTasks(Reconciliation reconciliation, String host, KafkaConnectApi apiClient, String connectorName, KafkaConnectorSpec connectorSpec, ConnectorStatusAndConditions status, CustomResource resource) {
+        JsonObject statusResultJson = new JsonObject(status.statusResult);
+        if (connectorSpec.getAutoRestart().isEnabled() && (connectorHasFailed(statusResultJson) || (failedTaskIds(statusResultJson)).size() > 0)) {
+            return getPreviousKafkaConnectorStatus(connectorOperator, resource)
+                 .compose(previousStatus -> {
+                     if (previousStatus == null) {
+                         previousStatus = new KafkaConnectorStatus();
+                     }
+                     return Future.succeededFuture(previousStatus.getAutoRestart());
+                 })
+                 .compose(previousAutoRestartStatus -> {
+                     if (shouldAutoRestart(previousAutoRestartStatus)) {
+                         LOGGER.debugCr(reconciliation, "Auto restarting connector {}", connectorName);
+                         return apiClient.restart(host, port, connectorName, true, true)
+                             .compose(
+                                 success -> Future.succeededFuture(status),
+                                 throwable -> {
+                                     status.autoRestart = StatusUtils.incrementAutoRestartStatus(previousAutoRestartStatus);
+                                     String message = "Failed to auto restart for the " + status.autoRestart.getCount() + " time(s) connector  " + connectorName + ". " + throwable.getMessage();
+                                     LOGGER.warnCr(reconciliation, message);
+                                     return Future.succeededFuture(status);
+                                 });
+                     } else {
+                         // keep previous auto restart status
+                         status.autoRestart = previousAutoRestartStatus;
+                         return Future.succeededFuture(status);
+                     }
+                 });
+        } else {
+            return Future.succeededFuture(status);
+        }
+    }
+
+    /**
+     * Connector or tasks failed should only restart at minute 0, 0, 2, 6, 12, 20 and 30
+     * @param autoRestartStatus
+     * @return true if the connector
+     */
+    private boolean shouldAutoRestart(AutoRestartStatus autoRestartStatus) {
+        if (autoRestartStatus == null) {
+            return true;
+        }
+        var count =  autoRestartStatus.getCount();
+        var minutesSinceLastRestart = StatusUtils.unitDifferenceUntilNow(StatusUtils.iso8601(autoRestartStatus.getLastRestartTimestamp()), ChronoUnit.MINUTES);
+
+        var nextRestart = count  * (count - 1);
+
+        return count <  7 &&  minutesSinceLastRestart  >= nextRestart;
```

Comment:
So, what happens after we restart it 7 times? Do we stop restarting it? I would expect it to restart it again after the longest interval (e.g. after 0, 2, 6, 12, 20, 30, 30, 30, 30 ... minutes0.)

Reply:
> between automatic restarts by 2 minutes, up to a maximum of 30 minutes, at which point no further automatic restarts are attempted, and the connector or tasks are left in the FAILED state.

I followed the proposal, should it keep restarting after 30 minutes ?

Reply:
Hmm, I missed this in the proposal. I guess it is good in that case. Although I'm not sure I understand the motivation behind it anymore.

Reply:
I guess the motivation would be that if the auto-restart fails 7 times we can probably assume that there is a problem with the config of the connector or task, so continuing to restart it forever will not help fix the problem.

Reply:
Ok, lets stick with it.

Reply:
Yes, that's right

> if the auto-restart fails 7 times we can probably assume that there is a problem with the config of the connector or task, so continuing to restart it forever will not help fix the problem.

Yes, that's the idea. The corollary here is that if the restart works and the connector is running happily for a while after restarting, but then fails again, it should start a new series of restarts. 

For example, I could imagine a connector that uses an external service, and that external service occasionally becomes unresponsive for a short time. We wouldn't want the connector to go into permanently failed state if it happens to have been restarted 7 times over a few days..